### PR TITLE
Support `osc new-app` from stored template

### PR DESCRIPTION
**Pull request by fabianofranz**

### CODE REVIEW

File: pkg/generate/app/cmd/newapp.go [hunk 1]
```diff
@@ -279,36 +292,68 @@ func (c *AppConfig) buildPipelines(components app.ComponentReferences, environme
...
-				strategy, source, err := app.StrategyAndSourceForRepository(ref.Input().Uses)
-				if err != nil {
-					return nil, fmt.Errorf("can't build %q: %v", ref.Input(), err)
-				}
-				if pipeline, err = app.NewBuildPipeline(ref.Input().String(), input, strategy, source); err != nil {
-					return nil, fmt.Errorf("can't build %q: %v", ref.Input(), err)
+			// TODO template and other individual matches are mutually exclusive,
+			// need to investigate the side-effects of having them together
+			if ref.Input().Match.Template != nil {
+				glog.V(2).Infof("will include objects from template %q", ref)
```
Comment (fabianofranz) [inline pkg/generate/app/cmd/newapp.go]: @smarterclayton I wasn't sure about the correct strategy here:

1. Break down the template and handle each of its objects as one individual pipeline here in `buildPipelines`. I think this is the right approach, but I will need to figure out how to handle Services, Routes, etc.
2. Break down the template before, maybe in `validate` and make each separate object a `ComponentInput` (I think it's wrong).
3. Handle the Template as one single pipeline element returned by `buildPipelines` and break it down later.

Thoughts?

Comment (smarterclayton) [inline pkg/generate/app/cmd/newapp.go]: Can you describe a scenario where it would make sense for a reasonably complex template to be paired with the input of other components?

I.e.:

    # mysql-cluster is a template
    $ osc new-app php~somegitrepo mysql-cluster
    # no overlap between components

    # mysql-cluster is a template, mysql is an image
    $ osc new-app mysql-cluster mysql
    # doesn't seem like the two should overlap, but they might conflict on their service name

The only case I can think of is when you have a template and a git repo:

    # eap is a template with a build config
    $ osc new-app <gitrepo> eap

In this case, the "eap" template would have to define a build config (or at least an input image) and we'd need to know to link those up.  So EAP's pipeline *might* be mutated to take the git repo as input... but that seems like a special case.

If the user specifies a single template, I don't think we should be breaking out the pipelines.  In the case above, it seems almost like a special case.  I would say that it's possible that templates should be given a chance to absorb arguments prior to execution, but then are added after the pipelines are joined.  Templates should probably be kept together as long as possible, but we should identify conflicts with the other arguments as soon as possible.

Honestly, at this point it might be better to disallow any mixing, and simply focus on getting the validation and checking behavior right.

----- Original Message -----

**Comment by fabianofranz** [comment]
This pull request adds the ability to create apps based on stored templates using `osc new-app`. The command will first process the stored template and then create based on it.

Provided that you first run `osc create -f examples/sample-app/application-template-stibuild.json` from the origin root dir to store the [stibuild template from sample-app](https://github.com/openshift/origin/blob/master/examples/sample-app/application-template-stibuild.json), here are a few usage examples.

Create based in the ruby-helloworld-sample template (currently 8 objects and 5 parameters).
```
osc new-app ruby-helloworld-sample
```

Same, overriding 2 of the 8 parameters specified by the template:
```
osc new-app ruby-helloworld-sample -e MYSQL_USER=myuser,MYSQL_PASSWORD=mypass
```

Different kinds of inputs *can* be mixed (templates, images, source code, etc); but at the moment the command *does not* accept templates with source code explicitly attached to it (e.g. `[template]~[code]`). Also, every input component is handled individually, we don't break the template to verify dependencies between components (e.g. source code was provided so the template should provide a build config). Also, object  names can conflict, we don't do prior validation against individual component names to check that (e.g. both the template and source code result in the creation of services with equal names).

```
osc new-app ruby-helloworld-sample mysql // OK
osc new-app ruby-helloworld-sample git@github.com/mfojtik/sinatra-app-example // OK, will create all components not attached to each other 
osc new-app ruby-helloworld-sample~git@github.com/mfojtik/sinatra-app-example // NOT SUPPORTED, the component before ~ can't be a template

**Comment by fabianofranz** [comment]
@smarterclayton How does it sound (comment above)? We still *could* make it more restrictive (e.g. simply not allow templates and source to be provided together), but since it doesn't break in most scenarios other than eventual naming conflicts, I'm allowing multiple kinds of inputs to be provided together as long as they are not explicitly attached to templates through the use of `~`.

**Comment by fabianofranz** [comment]
[test]

**Comment by fabianofranz** [comment]
@csrwng please review

**Comment by fabianofranz** [comment]
@csrwng All comments addressed. I ended up not using the same flag name `--value` used in `osc process` because in that command we are already in the context of manipulating templates, so "values" already implies "values to be set in the template". Here we are more generic so we have to be more specific in flag names.
Thanks for the review!

**Comment by smarterclayton** [comment]
Hold on merge of this until I have a chance to review tomorrow

**Comment by fabianofranz** [comment]
@smarterclayton ok, thanks. already accepted by QE.

File: hack/test-cmd.sh [hunk 2]
```diff
@@ -343,6 +343,16 @@ echo "imageRepositoryMappings: ok"
...
 osc new-app php mysql
+# check if we can create from a stored template
+osc create -f examples/sample-app/application-template-stibuild.json
+osc get template ruby-helloworld-sample
+[ "$(osc new-app ruby-helloworld-sample -o yaml | grep MYSQL_USER)" ]
+[ "$(osc new-app ruby-helloworld-sample -o yaml | grep MYSQL_PASSWORD)" ]
+[ "$(osc new-app ruby-helloworld-sample -o yaml | grep ADMIN_USERNAME)" ]
+[ "$(osc new-app ruby-helloworld-sample -o yaml | grep ADMIN_PASSWORD)" ]
+# create from template with code explicitly set is not supported
+[ ! "$(osc new-app ruby-helloworld-sample~git@github.com/mfojtik/sinatra-app-example)" ]
```
Comment (mfojtik) [inline hack/test-cmd.sh]: you should not use my repo ;-) I managed to break several times as I used it for my testing or testing features in STI :D

Comment (fabianofranz) [inline hack/test-cmd.sh]: It's a negative test (command will fail in validations), so the repo is there just for your own fame and profit.  :)

Comment (mfojtik) [inline hack/test-cmd.sh]: :+1:

Comment (smarterclayton) [inline hack/test-cmd.sh]: Actually we use it in osc help.  We need a good ootb example we can use in the docs - what's a good one?

File: pkg/cmd/cli/cmd/newapp.go [hunk 3]
```diff
@@ -30,9 +31,9 @@ var errExit = fmt.Errorf("exit directly")
 const newAppLongDesc = `
 Create a new application in OpenShift by specifying source code, templates, and/or images.
 
-This command will try to build up the components of an application using images or code
-located on your system. It will lookup the images on the local Docker installation (if
-available), a Docker registry, or an OpenShift image stream. If you specify a source
+This command will try to build up the components of an application using images, templates
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/newapp.go]: You need the Oxford comma (a, b, and c)

File: pkg/cmd/cli/cmd/newapp.go [hunk 4]
```diff
@@ -170,6 +176,16 @@ func RunNewApplication(f *clientcmd.Factory, out io.Writer, c *cobra.Command, ar
 				hasMissingRepo = true
 				fmt.Fprintf(c.Out(), "WARNING: We created an image stream %q, but it does not look like a Docker registry has been integrated with the OpenShift server. Automatic builds and deployments depend on that integration to detect new images and will not function properly.\n", t.Name)
 			}
+		case *deployapi.DeploymentConfig:
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/newapp.go]: Why did you add this?

File: pkg/cmd/cli/cmd/newapp.go [hunk 5]
```diff
@@ -80,10 +84,12 @@ func NewCmdNewApplication(fullName string, f *clientcmd.Factory, out io.Writer)
 	cmd.Flags().Var(&config.SourceRepositories, "code", "Source code to use to build this application.")
 	cmd.Flags().VarP(&config.ImageStreams, "image", "i", "Name of an OpenShift image stream to use in the app.")
 	cmd.Flags().Var(&config.DockerImages, "docker-image", "Name of a Docker image to include in the app.")
+	cmd.Flags().Var(&config.Templates, "stored-template", "Name of an OpenShift stored template to use in the app.")
+	cmd.Flags().Var(&config.TemplatesParameters, "stored-template-parameters", "Specify a list of key value pairs (eg. --stored-template-parameters=FOO=BAR,BAR=FOO) to set/override parameter values in the stored template.")
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/newapp.go]: These names are too long.  `--template` should be used (and the output version should be renamed `--output-template` for this command, or for now, just dropped).

`--stored-template-parameters` is about 20 characters too long.  I think this should be `--param` and `-p`

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/newapp.go]: Also "TemplateParameters", not "TemplatesParameters"

File: pkg/generate/app/cmd/newapp.go [hunk 6]
```diff
@@ -83,6 +85,11 @@ func (c *AppConfig) SetOpenShiftClient(osclient client.Interface, originNamespac
 		ImageStreamImages: osclient,
 		Namespaces:        []string{originNamespace, "default"},
 	}
+	c.templateResolver = app.TemplateResolver{
+		Client: osclient,
+		TemplateConfigsNamespacer: osclient,
+		Namespaces:                []string{originNamespace, "default", "openshift"},
```
Comment (smarterclayton) [inline pkg/generate/app/cmd/newapp.go]: originNamespace, "openshift", "default"

File: pkg/generate/app/cmd/newapp.go [hunk 7]
```diff
@@ -283,37 +307,36 @@ func (c *AppConfig) buildPipelines(components app.ComponentReferences, environme
...
-				if pipeline, err = app.NewBuildPipeline(ref.Input().String(), input, strategy, source); err != nil {
-					return nil, fmt.Errorf("can't build %q: %v", ref.Input(), err)
-				}
-
-			} else {
-				glog.V(2).Infof("will include %q", ref)
-				input, err := app.InputImageFromMatch(ref.Input().Match)
-				if err != nil {
-					return nil, fmt.Errorf("can't include %q: %v", ref.Input(), err)
+			if ref.Input().Match.IsImage() {
```
Comment (smarterclayton) [inline pkg/generate/app/cmd/newapp.go]: Rather than making this deeply nested, do 

```
if !ref.Input().Match.IsImage() {
  continue
}
...
```

File: pkg/generate/app/cmd/newapp.go [hunk 8]
```diff
@@ -386,6 +425,19 @@ func (c *AppConfig) Run(out io.Writer) (*AppResult, error) {
 
 	objects = app.AddServices(objects)
 
+	templates, err := c.buildTemplates(components, app.Environment(parameters))
+	if err != nil {
+		return nil, err
+	}
+	for _, t := range templates {
+		templateResolver := c.templateResolver.(app.TemplateResolver)
+		items, err := t.Process(templateResolver.TemplateConfigsNamespacer, templateResolver.Namespaces[0])
```
Comment (smarterclayton) [inline pkg/generate/app/cmd/newapp.go]: This is breaking encapsulation of process and the resolver, and doing buildTemplates up above but calling process here doesn't make this easier to read.  buildTemplates should include this entire block

File: pkg/generate/app/template.go [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	"github.com/openshift/origin/pkg/template"
+	templateapi "github.com/openshift/origin/pkg/template/api"
+)
+
+type Template struct {
+	From   string
+	Stored bool
+
+	Template *templateapi.Template
+	Env      Environment
```
Comment (smarterclayton) [inline pkg/generate/app/template.go]: Why is Env part of Template?  You should just pass it to Process

File: pkg/generate/app/templatelookup.go [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+type TemplateResolver struct {
+	Client                    client.TemplatesNamespacer
+	TemplateConfigsNamespacer client.TemplateConfigsNamespacer
+	Namespaces                []string
+}
+
+func (r TemplateResolver) Resolve(value string) (*ComponentMatch, error) {
+	namespaces := r.Namespaces
+	for _, namespace := range namespaces {
```
Comment (smarterclayton) [inline pkg/generate/app/templatelookup.go]: Don't check the same namespace twice (use a StringSet to check for that).

File: pkg/generate/app/template.go [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+func NewTemplate(from string, t *templateapi.Template, environment Environment) (*Template, error) {
+	return &Template{
+		From:     from,
+		Stored:   true,
+		Template: t,
+		Env:      environment,
+	}, nil
+}
+
+func (t *Template) Process(c client.TemplateConfigsNamespacer, namespace string) ([]runtime.Object, error) {
```
Comment (smarterclayton) [inline pkg/generate/app/template.go]: This whole class has very little value.  It would be better represented just as a single function.

File: pkg/generate/app/cmd/newapp_test.go [hunk 12]
```diff
@@ -147,5 +149,14 @@ func TestValidate(t *testing.T) {
...
 		}
+		if len(parms) != len(c.parms) {
+			t.Errorf("%s: Template parameters don't match. Expected: %v, Got: %v", n, c.parms, parms)
+		}
+		for p, v := range parms {
+			if c.parms[p] != v {
+				t.Errorf("%s: Template parameters don't match. Expected: %v, Got: %v", n, c.parms, parms)
+				break
+			}
+		}
```
Comment (smarterclayton) [inline pkg/generate/app/cmd/newapp_test.go]: add some tests for buildTemplate.

File: pkg/generate/app/cmd/newapp.go (hunk 8 above)
Comment (fabianofranz) [inline pkg/generate/app/cmd/newapp.go]: You got the "found breaking encapsulation piece" trophy. I knew you could do it people, was just waiting who was going to find it first! :)

Comment (smarterclayton) [inline pkg/generate/app/cmd/newapp.go]: :) yeah, it's ok to get info from the resolver but it's probably better to keep it distinct

File: pkg/cmd/cli/cmd/newapp.go (hunk 4 above)
Comment (fabianofranz) [inline pkg/cmd/cli/cmd/newapp.go]: The major reason was to print the generated parameter values (`MYSQL_USER`, `MYSQL_PASSWORD`, etc).

File: pkg/generate/app/cmd/newapp.go (hunk 8 above)
Comment (fabianofranz) [inline pkg/generate/app/cmd/newapp.go]: Is it ok to take the client from the resolver or should I just store it in `AppConfig`? I need it to be able to `Process`.

File: pkg/cmd/cli/cmd/newapp.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/newapp.go]: Hrm - I don't know that we need them here, especially if it's easy to get that info elsewhere.  Let's omit this for now and focus on making that easy in other places for now (ie describe dc/foo would show it).

File: pkg/generate/app/cmd/newapp.go (hunk 8 above)
Comment (smarterclayton) [inline pkg/generate/app/cmd/newapp.go]: AppConfig.  Be sure to take client.Interface

**Comment by fabianofranz** [comment]
@smarterclayton all comments addressed.

**Comment by smarterclayton** [comment]
One comment, then squash and I'll merge

**Comment by fabianofranz** [comment]
@smarterclayton fixed and squashed.

**Comment by smarterclayton** [comment]
LGTM [merge]
